### ISSUE TITLE
Intermittent "401 Unauthorized" failures from Chef server using Chef client

### ISSUE THREAD

**Issue by sds** [chef/chef-server#968]
We're seeing intermittent "401 Unauthorized" responses from the open source Chef server. These appear to be random events, but seem to occur when performing data bag searches (though that is by far our most common type of search, so other types may also have issues).

We've been having this issue for a long time through multiple Chef server and client upgrades. Our current versions are:
* Chef server: 12.9.1
* Chef client: 12.14.89

Before you dismiss this as a duplicate of #5, this is happening with the normal Chef client, *not* the `chef-api` gem.

Looking at the `crash.log` file for `opscode-erchef` doesn't reveal any additional information:
```bash
$ grep status=401 crash.log -B1
2016-10-13 17:43:32 =ERROR REPORT====
{<<"method=GET; path=/organizations/brigade/data/datadog/credentials; status=401; ">>,"Unauthorized"}
--
2016-10-13 17:47:57 =ERROR REPORT====
{<<"method=GET; path=/organizations/brigade/data/users/ambari; status=401; ">>,"Unauthorized"}
--
2016-10-13 18:08:08 =ERROR REPORT====
{<<"method=GET; path=/organizations/brigade/data/users/beanstalkd; status=401; ">>,"Unauthorized"}
--
2016-10-13 18:38:21 =ERROR REPORT====
{<<"method=GET; path=/organizations/brigade/data/users/gerrit; status=401; ">>,"Unauthorized"}
2016-10-13 18:38:21 =ERROR REPORT====
{<<"method=GET; path=/organizations/brigade/data/users/kibanna; status=401; ">>,"Unauthorized"}
2016-10-13 18:38:21 =ERROR REPORT====
{<<"method=GET; path=/organizations/brigade/data/users/solr; status=401; ">>,"Unauthorized"}
```

Here's an example of the stack trace on the Chef client side (though less useful since the problem is server-side):
```
Net::HTTPServerException: 401 "Unauthorized"
  from net/http/response.rb:120:in `error!'
  from chef/http.rb:150:in `request'
  from chef/http.rb:115:in `get'
  from chef/data_bag_item.rb:155:in `load'
  from chef/dsl/data_query.rb:62:in `data_bag_item'
  from /var/chef/cache/cookbooks/base/recipes/_sudo.rb:7:in `block in from_file'
  from /var/chef/cache/cookbooks/base/recipes/_sudo.rb:6:in `each'
  from /var/chef/cache/cookbooks/base/recipes/_sudo.rb:6:in `from_file'
  from chef/mixin/from_file.rb:30:in `instance_eval'
  from chef/mixin/from_file.rb:30:in `from_file'
  from chef/cookbook_version.rb:233:in `load_recipe'
  from chef/run_context.rb:327:in `load_recipe'
  from chef/run_context.rb:283:in `block in include_recipe'
  from chef/run_context.rb:282:in `each'
  from chef/run_context.rb:282:in `include_recipe'
  from chef/dsl/include_recipe.rb:26:in `include_recipe'
  from /var/chef/cache/cookbooks/base/recipes/default.rb:23:in `from_file'
  from chef/mixin/from_file.rb:30:in `instance_eval'
  from chef/mixin/from_file.rb:30:in `from_file'
  from chef/cookbook_version.rb:233:in `load_recipe'
  from chef/run_context.rb:327:in `load_recipe'
  from chef/run_context/cookbook_compiler.rb:141:in `block in compile_recipes'
  from chef/run_context/cookbook_compiler.rb:138:in `each'
  from chef/run_context/cookbook_compiler.rb:138:in `compile_recipes'
  from chef/run_context/cookbook_compiler.rb:75:in `compile'
  from chef/run_context.rb:176:in `load'
  from chef/policy_builder/expand_node_object.rb:97:in `setup_run_context'
  from chef/client.rb:510:in `setup_run_context'
  from chef/client.rb:280:in `run'
  from chef/application.rb:302:in `block in fork_chef_client'
  from chef/application.rb:290:in `fork'
  from chef/application.rb:290:in `fork_chef_client'
  from chef/application.rb:255:in `block in run_chef_client'
  from chef/local_mode.rb:44:in `with_server_connectivity'
  from chef/application.rb:243:in `run_chef_client'
  from chef/application/client.rb:464:in `sleep_then_run_chef_client'
  from chef/application/client.rb:451:in `block in interval_run_chef_client'
  from chef/application/client.rb:450:in `loop'
  from chef/application/client.rb:450:in `interval_run_chef_client'
  from chef/application/client.rb:434:in `run_application'
  from chef/application.rb:60:in `run'
  from /opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.14.89/bin/chef-client:26:in `<top (required)>'
  from /usr/bin/chef-client:54:in `load'
  from /usr/bin/chef-client:54:in `<main>'
```

We've intentionally rebuilt our entire Chef server from scratch with the hope that this would fix the problem, to no avail.

What additional information can we provide to help you debug this issue?

**Comment by marcparadise** [comment]
Do the same requests fail consistently from the same clients/users?

**Comment by sds** [comment]
No, there does not appear to be a clear pattern here. I just checked our Sentry logs, and there's a long tail of nodes that account for 2-5% of errors in the last 7 days. No obvious pattern there.

**Comment by marcparadise** [comment]
Okay, thanks. Could you post a segment of hte chef-server request log that shows a couple-few of these? You can find it in `/var/opt/chef-server/opscode-erchef/request*`

**Comment by sds** [comment]
Happy to! Do you mind if I email you a link to a secret gist? Would prefer to not post information in the open, if possible.

**Comment by sds** [comment]
Sent you a private gist to your email. Let me know if there's anything else I can provide!

**Comment by marcparadise** [comment]
Thanks, I got that.  

The first thing that jumps out is that the requests are failing with the message `bad_clock` - that is intended to mean "your chef client provided a request timestamp that is too old" - 900s or more behind current server time.

That seems like it would produce a lot more of these errors though, though hovering  right on that 900s border would explain the intermittent nature.  I don't thikn this is the case, but could you verify that the nodes are in sync time-wise with the chef server?

Beyond that, would it be possible to enable chef client debug logging on a few nodes? It would give us a lot more information to work with if we could see the exact request that chef-client is sending.

Final question - do your logs show any occurrences of this in endpoints other that data bags or search?

**Comment by sds** [comment]
One quick thing I've noticed which I didn't before: all of the nodes which experience this problem are in a specific Chef environment (`production`). I apologize for not noticing this before. This suggests some sort of difference in configuration that is reproducible, so I'm going to investigate that separately (though I'm pretty sure we don't have different Chef client configurations between production and our other environments).

In regards to your questions:

Our servers clocks are all synced with NTP, but I verified by comparing each node's clock against the Chef server's clock:

```bash
knife ssh "*:*" 'expr $(date +%s) - $(ssh chef.brigade.com date +%s)' -C32
```

None of the hosts had more than a couple of seconds of clock skew from the Chef server. 

I've also verified that they are all running in the same time zone (UTC):

```bash
knife ssh "*:*" 'date +%Z' -C32
```

I'll turn on debug-level logging for a sizable subset of our production nodes so that when this happens again we can hopefully catch the problematic request.

*All* the 401s we see are the result of data bag searches (although as I mentioned above, that is the most common kind of search we use by far, so it's possible that the problem may not be specific to data bags, but searches of all kinds). The logs unfortunately don't go back far enough for me to be certain, but the evidence we have for the last week is compelling.

Thanks for your help so far! I'll keep digging in to what's different about our production nodes, and report back if we can catch the logs on the client side when this error occurs.

**Comment by sds** [comment]
False alarm with respect to this only occurring in the production Chef environment: turns out we were only reporting errors to Sentry for the production environment, hence the discrepancy.

I'll report back once we've got logs for a request on the Chef client side.

**Comment by sds** [comment]
Ok, I see what's happening here.

Upon further investigation: these errors do not occur with perfectly random distribution. They appeared to, but this was for a reason we'll see shortly.

We backup our Chef server on a daily basis by running `chef-server-ctl backup` on the server. This requires a short period of downtime during which the server is unavailable. This is fine, since the Chef client retries HTTP requests when the server is unreachable. 

Here is a relevant section of logs where a 401 Unauthorized request was returned:

```
[2016-10-15T17:00:16+00:00] ERROR: Connection refused connecting to https://chef.brigade.com/organizations/brigade/data/certificates/staging, retry 2/12
[2016-10-15T17:00:21+00:00] ERROR: Connection refused connecting to https://chef.brigade.com/organizations/brigade/data/certificates/staging, retry 3/12
[2016-10-15T17:00:26+00:00] ERROR: Connection refused connecting to https://chef.brigade.com/organizations/brigade/data/certificates/staging, retry 4/12
[2016-10-15T17:00:31+00:00] ERROR: Connection refused connecting to https://chef.brigade.com/organizations/brigade/data/certificates/staging, retry 5/12
[2016-10-15T17:00:36+00:00] ERROR: Connection refused connecting to https://chef.brigade.com/organizations/brigade/data/certificates/staging, retry 6/12
[2016-10-15T17:00:41+00:00] ERROR: Connection refused connecting to https://chef.brigade.com/organizations/brigade/data/certificates/staging, retry 7/12
[2016-10-15T17:00:46+00:00] ERROR: Connection refused connecting to https://chef.brigade.com/organizations/brigade/data/certificates/staging, retry 8/12
[2016-10-15T17:00:51+00:00] ERROR: Connection refused connecting to https://chef.brigade.com/organizations/brigade/data/certificates/staging, retry 9/12
[2016-10-15T17:00:56+00:00] ERROR: Server returned error 503 for https://chef.brigade.com/organizations/brigade/data/certificates/staging, retrying 10/12 in 1732s
[2016-10-15T17:29:48+00:00] INFO: HTTP Request Returned 401 Unauthorized: error
certificate_pem[/config/staging.pem] action create (1777.148678337s) FAILED: 401 "Unauthorized"
```

Here we see for the first 9 tries, the Chef client can't even open a connection due to the downtime caused by the backup. The [current code](https://github.com/chef/chef/blob/v12.14.89/lib/chef/http.rb#L393-L398) has it retrying every `http_retry_delay` seconds in this case ([5 by default](https://github.com/chef/chef/blob/v12.14.89/chef-config/lib/chef-config/config.rb#L314)) up to `http_retry_count` times (5 by default, but we've set ours to 12 because of the backup downtime), which is why we get up to 9 tries rather quickly:

```ruby
      rescue Errno::ECONNREFUSED
        if http_retry_count - http_attempts + 1 > 0
          Chef::Log.error("Connection refused connecting to #{url}, retry #{http_attempts}/#{http_retry_count}")
          sleep(http_retry_delay)
          retry
        end
```

Unfortunately, this means that once the nginx server comes back up, it's possible that other backend services on the server are not up yet, so while the connection is opened the request itself will 503, whose retry logic is [handled differently](https://github.com/chef/chef/blob/v12.14.89/lib/chef/http.rb#L375-L382):

```ruby
          if response.kind_of?(Net::HTTPServerError) && !Chef::Config.local_mode
            if http_retry_count - http_attempts + 1 > 0
              sleep_time = 1 + (2**http_attempts) + rand(2**http_attempts)
              Chef::Log.error("Server returned error #{response.code} for #{url}, retrying #{http_attempts}/#{http_retry_count} in #{sleep_time}s")
              sleep(sleep_time)
              redo
            end
          end
```

Since this is the 10th try, we now sleep 2^10 = 1024 seconds + a random range between 1 and 1024. **This is why these events appeared random**, as the retried request could have happened anywhere from 17 to 34 minutes, or an even wider range depending on the number of times we failed to open the initial connection.

The reason the final request 401s is because the authorization headers are generated at the [very beginning of this retry process](https://github.com/chef/chef/blob/v12.14.89/lib/chef/http.rb#L333), with the same headers used on each retry. If you exceed 900 seconds of retry delay you'll have headers that are out-of-date.

In summary: this is not a problem with the Chef server, but the Chef client. I'll open a pull request with the Chef client project, as from my cursory understanding we can easily avoid this problem by regenerating the headers before each retry. There may be a technical reason why they are generated once at the beginning, but I'll open a pull request and see what you think.

Thanks for you help on this, @marcparadise!

**Comment by sds** [comment]
I should add that a simple workaround if this pull request isn't acceptable is to increase your `http_retry_delay` so that you rack up fewer retries before hitting the 503.

**Comment by marcparadise** [comment]
@sds thanks for continuing to dig in, and for the PR - good find!

**Comment by sds** [comment]
Closing since the pull request was merged in the Chef client project.
